porch/pkg/engine/engine.go

```diff
@@ -515,23 +521,33 @@ func (m *updatePackageMutation) Apply(ctx context.Context, resources repository.
 	if err != nil {
 		klog.Infof("failed to add merge key comments: %v", err)
 	}
-	return result, m.task, nil
+	return result, m.updateTask, nil
 }
 
 // Currently assumption is that downstream packages will be forked from a porch package.
-// As per current implementation, upstream package ref is stored in an old clone task, but this may
+// As per current implementation, upstream package ref is stored in a new update task but this may
 // change so the logic of figuring out current upstream will live in this function.
 func (m *updatePackageMutation) currUpstream() (*api.PackageRevisionRef, error) {
-	if m.oldTask == nil || m.oldTask.Clone == nil {
-		return nil, fmt.Errorf("package %s does not have upstream info", m.pkgName)
+	if m.cloneTask == nil || m.cloneTask.Clone == nil {
+		return nil, fmt.Errorf("package %s does not have original upstream info", m.pkgName)
 	}
-	upstream := m.oldTask.Clone.Upstream
+	upstream := m.cloneTask.Clone.Upstream
 	if upstream.Type == api.RepositoryTypeGit || upstream.Type == api.RepositoryTypeOCI {
 		return nil, fmt.Errorf("upstream package must be porch native package. Found it to be %s", upstream.Type)
 	}
 	return upstream.UpstreamRef, nil
 }
 
+func findCloneTask(pr *api.PackageRevision) *api.Task {
```

Comment:
With this change, a clone task can only exist as the first task in the list. Every packagerevision must start with either an InitTask or a CloneTask

Reply:
Fixed